ocaml/xapi/xapi_xenops.ml

```diff
@@ -2430,6 +2432,7 @@ let start ~__context ~self paused force =
       let module Client = (val make_client queue_name : XENOPS) in
       debug "Sending VM %s configuration to xenopsd" (Ref.string_of self);
       try
+        if not force then assert_memory_constraints ~__context ~vm:self;
```

Comment:
It may be possible, and better, to do this assertion a little earlier, because this point we have already done a couple of things, such as setting `currently_attached` to `true` on VIFs and VBDs (which is not reverted if the assertion failed).

Since it is just a DB-based check, it can be done right in the beginning of `start` in xapi_vm.ml (there is already a `Vm_platform.check_restricted_flags` check in there, which is a similar kind of thing).